### ISSUE TITLE
Testing CSS in styled subcomponent

### ISSUE THREAD

**Issue by dariobanfi** [styled-components/jest-styled-components#77]
Hello,

I recently upgraded to React 16 and Enzyme 3.0.0 and jest-styled-components 4.6.0, and it seems this type of test, which before worked, donesn't work anymore:

```
const C1 = styled.div`
  color: 'something';
`
const C2 = styled(C1)`
  padding: 0;
`
...

expect(wrapper.find(C2)).toHaveStyleRule('color', 'something');
```

C2 has only its own rules and not the inherited ones.
Is this an intended behaviour? Any suggestions?

Thanks in advance for the awesome library!

**Comment by MicheleBertoli** [comment]
Hello @dariobanfi, thanks for opening the issue.
Would you be able to provide a full non-working example (e.g. including the rendering part)?
Also, which version of this library were you using before?
Thank you very much!

**Comment by doque** [comment]
Hi @MicheleBertoli , I work with @dariobanfi on this and have done some further investigating.
The underlying issue seems to be that calling `styled(Component)` (as opposed to `Component.extend`) causes the style rules of the root components to be dropped.

An example repository can be found [here](https://github.com/doque/jest-styled-components-properties/blob/master/src/App.test.js#L20) (run `npm test` run the test suite).

This is the test:
```javascript
import styled, { ThemeProvider } from 'styled-components';
import { mount } from 'enzyme';


const Parent = styled.div`
  border-color: ${props =>
    props.one ? props.theme.colors.one : props.theme.colors.two};
  background-color: ${props => props.theme.colors.three };
`;

const Child = styled(Parent)`
  background-color: ${props => props.theme.colors.four };
`;

describe("Theme is applied correctly", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <ThemeProvider theme={myTheme}>
        <div>
          <Parent one />
          <Child one/>
        </div>
      </ThemeProvider>
    );
  });

  test("Parent has the correct border-color", () => {
    expect(wrapper.find(Parent)).toHaveStyleRule(
      "border-color",
      myTheme.colors.one
    );
  });

  test("Child has the correct background-color", () => {
    expect(wrapper.find(Child)).toHaveStyleRule("background-color", myTheme.colors.four);
  })

  // This test breaks, border-color property not found in Child
  test("Child has the correct inherited border-color", () => {
    expect(wrapper.find(Child)).toHaveStyleRule("border-color", myTheme.colors.one);
  })

});
```
The last test breaks. It will, however, pass if the definition of `Child` is changed to `const Child = Parent.extend` instead of `const Child = styled(Parent)`.

Oddly, all styles are in fact applied correctly to the component - it looks the way it should. But in the tests, the style rules defined for `Parent` are not present for `Child`.

As far as I know, the only difference between `extend` and `styled()` is the className generation, according to the styled-components docs:
![image](https://user-images.githubusercontent.com/1480751/31169735-e63e586a-a8f9-11e7-9242-d4f6b749dfc7.png)

Are we missing something here? Thanks for your help!


For reference, these are the dependency versions we are using/were using before the upgrade (the `jest-styled-components` version hasn't changed):
```diff
+    "enzyme": "3.0.0",
+    "enzyme-adapter-react-16": "1.0.0",
+    "enzyme-to-json": "3.0.1",
-    "jest": "21.0.2",
+    "jest": "21.2.0",
     "jest-styled-components": "4.6.0",
-    "react-addons-test-utils": "15.6.0",
+    "react-addons-test-utils": "15.6.2",

```

**Comment by MicheleBertoli** [comment]
Thanks @doque @dariobanfi for providing more information, this is a really interesting issue.

I created a small test case to repro the issue:

```js
const Parent = styled.div`
  color: red;
`

const StyledChild = styled(Parent)`
  padding: 0;
`

const ExtendChild = Parent.extend`
  padding: 0;
`

test('styled / shallow', () => {
  const wrapper = shallow(<StyledChild  />)
  console.log(wrapper.debug())
  expect(wrapper).toHaveStyleRule('color', 'red')
})

// FAILS
// <styled.div className="sc-bwzfXH iRycfs" innerRef={[undefined]} />

test('styled / mount', () => {
  const wrapper = mount(<StyledChild  />)
  console.log(wrapper.debug())
  expect(wrapper).toHaveStyleRule('color', 'red')
})

// FAILS
// <Styled(styled.div)>
//   <styled.div className="sc-bwzfXH iRycfs" innerRef={[undefined]}>
//     <div className="sc-bwzfXH iRycfs sc-bdVaJa rOCEJ" />
//   </styled.div>
// </Styled(styled.div)>

test('extend / shallow', () => {
  const wrapper = shallow(<ExtendChild  />)
  console.log(wrapper.debug())
  expect(wrapper).toHaveStyleRule('color', 'red')
})

// PASSES
// <div className="sc-htpNat kMmqyg" />

test('extend / mount', () => {
  const wrapper = mount(<ExtendChild  />)
  console.log(wrapper.debug())
  expect(wrapper).toHaveStyleRule('color', 'red')
})

// PASSES
// <styled.div>
//   <div className="sc-htpNat kMmqyg" />
// </styled.div>

```

The root cause is that when using `styled`, Enzyme wraps the actual element with classes `sc-bwzfXH iRycfs sc-bdVaJa rOCEJ` (that contains the style rules for both parent and child) into a `styled.div` with classes `sc-bwzfXH iRycfs` (that contains the style rules for the child only).
With `extend` instead, as per documentation, a single class name is generated.

We should find a way to bypass the middle element, and check the rules on the innermost child.

**Comment by MicheleBertoli** [comment]
I made an [experiment](https://github.com/styled-components/jest-styled-components/compare/styled-child), and it seems to work.
If you want to give it a go:`yarn add --dev styled-components@next`.
Any feedback is much appreciated, thanks!

**Comment by MicheleBertoli** [comment]
Fixed in [v4.9.0](https://github.com/styled-components/jest-styled-components/blob/master/CHANGELOG.md#490---2017-10-22), thank you very much!
